torch/csrc/jit/python/python_dict.cpp

```diff
@@ -0,0 +1,198 @@
+#include <ATen/core/ivalue.h>
+#include <pybind11/cast.h>
+#include <pybind11/detail/common.h>
+#include <torch/csrc/jit/python/pybind_utils.h>
+#include <torch/csrc/jit/python/python_dict.h>
+#include <torch/csrc/jit/runtime/jit_exception.h>
+#include <sstream>
+#include <stdexcept>
+
+namespace torch {
+namespace jit {
+
+struct CustomMethodProxy;
+struct CustomObjectProxy;
+
+IValue ScriptDictIterator::next() {
+  if (iter_ == end_) {
+    throw py::stop_iteration();
+  }
+
+  // Since this is the iterator for .items(), the current key and value
+  // should be returned as a tuple.
+  IValue result = c10::ivalue::Tuple::create({iter_->key(), iter_->value()});
+
+  // Advance the iterator for next time.
+  iter_++;
+
+  return result;
+}
+
+IValue ScriptDictKeyIterator::next() {
+  if (iter_ == end_) {
+    throw py::stop_iteration();
+  }
+
+  // Since this is the iterator for .keys() and __iter__(), return only the key.
+  IValue result = iter_->key();
+
+  // Advance the iterator for next time.
+  iter_++;
+
+  return result;
+}
+
+void initScriptDictBindings(PyObject* module) {
+  auto m = py::handle(module).cast<py::module>();
+
+  py::class_<ScriptDictKeyIterator>(m, "ScriptDictKeyIterator")
+      .def(
+          "__next__",
+          [](ScriptDictKeyIterator& iter) {
+            auto result = iter.next();
+            return toPyObject(result);
+          })
+      .def("__iter__", [](ScriptDictKeyIterator& iter) { return iter; });
+
+  py::class_<ScriptDictIterator>(m, "ScriptDictIterator")
+      .def(
+          "__next__",
+          [](ScriptDictIterator& iter) {
+            auto result = iter.next();
+            return toPyObject(result);
+          })
+      .def("__iter__", [](ScriptDictIterator& iter) { return iter; });
+
+  py::class_<ScriptDict, std::shared_ptr<ScriptDict>>(m, "ScriptDict")
+      .def(py::init([](py::dict dict) {
+        TypePtr type = nullptr;
+
+        if (dict.size() > 0) {
+          // If the source dictionary is nonempty, try to infer its type.
+          auto inferred_type = tryToInferType(dict);
+
+          if (!inferred_type.success()) {
+            std::stringstream ss;
+            ss << "Unable to infer type of dictionary: "
+               << inferred_type.reason();
+            throw JITException(ss.str());
+          }
+
+          type = inferred_type.type();
+        } else {
+          // If is empty, assume the type is Dict[str, Tensor] as is done in
+          // TorchScript code.
+          type = DictType::create(StringType::get(), TensorType::getInferred());
```

Comment:
@ansley Note that this is another place you might want to pay attention to when inferring type for `{}` and `[]`